### PR TITLE
SIP273 Add waterDrainFrac to allow water > WHC

### LINKED ISSUE #273: Add waterDrainFrac to allow flooding.
Currently, SIPNET drains all soil water in excess of water holding capacity (WHC) instantly. To allow realistic simulation of saturated or flooded soil conditions, our chosen option is to add a parameter that controls drainage rate.

Add a parameter (e.g., `waterDrainFrac` ($f_{\text{drain}}$ in documentation) that controls the fraction of excess soil water drained per day:

Setting `waterDrainFrac < 1` would enable prolonged saturated/flooded conditions. This will enable realistic representation of managed flooding (e.g., rice), as well as poorly drained / wetland ecosystems.

### PR DESCRIPTION
**Pull request by Alomir**
## Summary

- **What**: Adds a control that allows soil water to go above WHC, enabling modeling of flooding
- **Motivation**: Necessary for project

## How was this change tested?
Added unit tests in `tests/sipnet/test_modeling/testSoilMoisture.c` covering:
- `getClippedWaterFrac`: boundary cases (within range, at WHC, above WHC, at zero, below zero)
- `calcSoilWaterFluxes` drainage with `waterDrainFrac` = 1.0, 0.5, 0.0, and no drainage when water ≤ WHC
- `moisture()`: transpiration is capped at soilWHC (not soilWater) when soil is flooded

```bash
make smoke
python tools/smoke_check.py run verbose
```
`smoke_check.py` results:
[sc.txt](https://github.com/user-attachments/files/26419168/sc.txt)
Minor differences in ET due to the added water frac cap

## Related issues

- Fixes #273

## Checklist

- [x] Related issues are listed above. [PRs without an approved, related issue may not get reviewed](docs/CONTRIBUTING.md#propose-and-receive-feedback).
- [x] PR title has the issue number in it ("[#<number>] \<concise description of proposed change>")
- [x] Tests added/updated for new features (if applicable)
- [x] Documentation updated (if applicable)
- [x] `docs/CHANGELOG.md` updated with noteworthy changes
- [x] Code formatted with `clang-format` (run `git clang-format` if needed)

### CODE REVIEW

**Comment by Alomir** [comment]
@copilot please add a test that covers these changes. The test should be named `testSoilMoisture.c` in the directory `tests/sipnet/test_modeling`, and it should be written in the style of other tests in that directory.

**Comment by Alomir** [comment]
@copilot please remove all changes in the tests/sipnet/test_restart_infrastucture directory that were added in [552adf5](https://github.com/PecanProject/sipnet/pull/307/commits/552adf539ba4781617caaa10299d892c526b1a58)

File: src/sipnet/sipnet.c [hunk 1]
```diff
@@ -977,7 +995,9 @@ void calcSoilWaterFluxes(double *fastFlow, double *evaporation,
 
   // drain any water that remains beyond water holding capacity:
   if (waterRemaining > params.soilWHC) {
-    *drainage = (waterRemaining - params.soilWHC) / (climate->length);
+    // waterDrainFrac is per-time-step, not per-day
+    *drainage = (waterRemaining - params.soilWHC) / (climate->length) *
+                params.waterDrainFrac;
```
[the remark this answers is not in the document]
Comment (Alomir) [reply src/sipnet/sipnet.c]: Ticket #278 will address this on a larger scale

File: src/sipnet/sipnet.c [hunk 2]
```diff
@@ -901,6 +906,23 @@ void snowPack(double *snowMelt, double *sublimation, double snowFall) {
   }  // end else there is snow
 }  // end snowPack
 
+/**
+ * Calculated water fraction bounded in [0,1]
```
[the remark this answers is not in the document]
Comment (Alomir) [reply src/sipnet/sipnet.c]: 👍

File: src/sipnet/sipnet.c [hunk 3]
```diff
@@ -385,6 +385,9 @@ void readParamData(ModelParams **modelParamsPtr, const char *paramFile) {
   initializeOneModelParam(modelParams, "soilMethaneRate", &(params.soilMethaneRate), ctx.anaerobic);
   initializeOneModelParam(modelParams, "litterMethaneRate", &(params.litterMethaneRate), ctx.anaerobic);
 
+  // Water drainage
+  initializeOneModelParam(modelParams, "waterDrainFrac", &(params.waterDrainFrac), 1);
```
[the remark this answers is not in the document]
Comment (Alomir) [reply src/sipnet/sipnet.c]: I hear the issue here - but this would be the first and only param to actually have a default value. I'll update the doc.

**Comment by Alomir** [comment]
Moving back to draft while I consider whether to change to a model-flag approach for backward compatibility

File: docs/model-structure.md
Comment (dlebauer) [inline docs/model-structure.md:878]:
```suggestion
F^W_\text{trans} = \min\(F^W_\text{trans, pot}, f \cdot \min\(W_\text{WHC}, W_\text{soil}\)\)
```

Should we define plant available soil water, `\min\(W_\text{WHC}, W_\text{soil}\)\)` separately for clarity? (update: deferred to #309)

Comment (dlebauer) [inline docs/model-structure.md]:
```suggestion
fraction  $(f)$ of the plant-available soil water  $(W_\text{soil})$ that is removable in one day.
Plant-available soil water is capped at the water holding capacity $W_\text{WHC}$.
```

Comment (dlebauer) [inline docs/model-structure.md:832]: for consistency w/ code 
not sure most clear notation: clip, min(1,max(0,x)), or otherwise

Or sufficient to explain in text?

```suggestion
r_d = \frac{\text{rdConst}}{u},
\qquad r_{\text{soil}} = \exp\!\left(r_{\text{soil},1} -
r_{\text{soil},2}\,
\operatorname{clip}\!\left(\frac{W_{\text{soil}}}{W_{\text{WHC}}}, 0, 1\right)
\right)
```

File: docs/parameters.md
Comment (dlebauer) [inline docs/parameters.md]: This is just to clarify:

```suggestion
| $W_{\text{soil},0}$               | soilWFracInit  | Initial soil water content                                               | unitless                                                            | Fraction of soilWHC, may be greater than 1 when initializing flooded conditions          |
```

But, I think we are overloading this term - and if we call it f_WHC we can clean up the math notation - many of the equations I just edited use $\frac{W_{\text{soil}}}{W_{\text{WHC}}}$ as the unitless fraction, and it will be more clear if we replace that w/ f_WHC.

But, I am going to separate out this docs cleanup into a separate issue. If you want to ignore some of the other docs changes and just link to that issue that is okay.

File: docs/parameters.md [hunk 4]
```diff
@@ -210,24 +210,25 @@ pool stoichiometry.
...
+| $r_{s,1}$                       | rSoilConst1         |                                                                                                                        | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1) , where W1 = (water/soilWHC)                                                                                                      |
+| $r_{s,2}$                       | rSoilConst2         |                                                                                                                        | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1) , where W1 = (water/soilWHC)                                                                                                      |
```
Comment (dlebauer) [inline docs/parameters.md]:
```suggestion
| $r_{s,1}$                       | rSoilConst1         |                                                                                                                        | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1), where W1 = clip(water/soilWHC, 0, 1)                                                                                            |
| $r_{s,2}$                       | rSoilConst2         |                                                                                                                        | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1), where W1 = clip(water/soilWHC, 0, 1)                                                                                            |
```

File: docs/model-structure.md [hunk 5]
```diff
@@ -738,11 +738,13 @@ irrigation) is assumed to evaporate the same day and therefore never enters $W_{
...
+holding capacity $(W_{\text{WHC}})$. Beyond this point, additional water drains off at a rate controlled by the
+drainage parameter $f_{\text{drain}}$, defined as the fraction of excess soil water that can be removed per time step.
+For well drained soils, set $f_{\text{drain}} = 1$ to ensure full drainage with each step.
+
+Setting $f_{\text{drain}} < 1$ reduces the rate of drainage. Flooding can be simulated by using a combination of low 
+$f_{\text{drain}}$ and sufficient $F^W_\text{irrig|precip,soil}$ to maintain flooded conditions. Note, though, that 
+if the model time steps are not constant, the actual per-day drainage rate will differ in the different steps.
 
 \begin{equation}
 F^W_{\text{drainage}} = f_\text{drain} \cdot \max(W_{\text{soil}} - W_{\text{WHC}}, 0)
```
Comment (dlebauer) [inline docs/model-structure.md:756]: Does this need 1 / \Delta t on rhs?


```suggestion
F^W_{\text{drainage}} = \frac{1}{\Delta t} f_\text{drain} \cdot \max(W_{\text{soil}} - W_{\text{WHC}}, 0)
```

File: docs/model-structure.md
Comment (dlebauer) [inline docs/model-structure.md:839]:
```suggestion

For soil evaporation, SIPNET assumes a saturated-surface approximation by clipping $\frac{W_\text{soil}}{W_\text{WHC}}$ to $[0,1]$ before calculating $r_\text{soil}$.

```

File: docs/parameters.md [hunk 6]
```diff
@@ -210,24 +210,25 @@ pool stoichiometry.
...
+| $f_{\text{fastflow}}$           | fastFlowFrac        | fraction of water entering soil that goes directly to drainage                                                         | unitless                                       |                                                                                                                                                                                         |
+| $f_{\text{drain}}$              | waterDrainFrac      | fraction of excess soil water that can be removed per time step                                                        | unitless                                       | Enables flooding when less than 1                                                                                                                                                       |
```
Comment (dlebauer) [inline docs/parameters.md]: to confirm - is this the only non-formatting change?

File: docs/parameters.md
Comment (dlebauer) [reply docs/parameters.md]: Addressed in #309 (but suggested change can still be accepted here)

**Review by dlebauer — CHANGES_REQUESTED** [review]
Some documentation updates to align w/ implementation. Please check correctness. Feel free to defer any suggestions to #309 (but I think  it is prudent to add the clipping and 1/delta t w/ this PR)

**Comment by Alomir** [comment]
@copilot - please revert all changes in the files listed below in this PR. Make no other changes. I understand that the tests may not pass after this.

src/common/context.c
src/common/context.h
src/sipnet/cli.c
src/sipnet/cli.h
src/sipnet/restart.c
tests/sipnet/test_restart_infrastructure/restart.param
tests/sipnet/test_restart_infrastructure/spatial_val.param
tests/sipnet/test_sipnet_infrastructure/standard.exp
tests/sipnet/test_sipnet_infrastructure/standard.param
tests/sipnet/test_sipnet_infrastructure/with_est.exp
tests/sipnet/test_sipnet_infrastructure/with_est.param
tests/smoke/niwot/sipnet.param
tests/smoke/niwot/sipnet.config
tests/smoke/russell_1/sipnet.config
tests/smoke/russell_1/sipnet.param
tests/smoke/russell_2/sipnet.config
tests/smoke/russell_2/sipnet.param
tests/smoke/russell_3/sipnet.config
tests/smoke/russell_3/sipnet.param

File: docs/model-structure.md (hunk 5 above)
Comment (Alomir) [reply docs/model-structure.md:756]: Yes!

File: docs/model-structure.md
Comment (Alomir) [reply docs/model-structure.md:832]: I went with the "explain in text" choice

File: docs/parameters.md (hunk 4 above)
Comment (Alomir) [reply docs/parameters.md]: 👍

File: docs/parameters.md (hunk 6 above)
Comment (Alomir) [reply docs/parameters.md]: Yes (other than your suggested changes)

File: src/common/modelParams.c [hunk 7]
```diff
@@ -272,3 +272,10 @@ void deleteModelParams(ModelParams *modelParams) {
   free(modelParams->readIndices);
   free(modelParams);
 }
+
+void setDefaultParamValue(ModelParams *params, char *pName, double value) {
+  int paramIndex = locateParam(params, pName);
+  OneModelParam *param = &(params->params[paramIndex]);
+  *(param->value) = value;
+  param->isRead = -1;
```
[the remark this answers is not in the document]
Comment (Alomir) [reply src/common/modelParams.c]: That's the intended behavior for default values, even though this is a bit hack-ish. If we get another param needing a default, we'll make this more official.

I'm fine with no checks in this function, as this is all internal.

File: docs/CHANGELOG.md [hunk 8]
```diff
@@ -50,6 +48,7 @@ sections to include in release notes:
 - Nitrogen demand, fixation flux, and uptake flux (#265)
 - Methane production option, controlled by the `anaerobic` cli option (#269)
 - Nitrogen limitation support (#298)
+- Parameter waterDrainFrac to enable flooded conditions. Note that this is a required param, so older files without it will not run. (#307)
```
[the remark this answers is not in the document]
Comment (Alomir) [reply docs/CHANGELOG.md]: Right, missed this!

**Review by dlebauer — APPROVED** [review]
LGTM, looking forward to some flooded 🌾 & 🍚!

File: src/sipnet/restart.c [hunk 9]
```diff
@@ -160,7 +161,8 @@ void initResetState(RestartState *state, MeanTracker *npp) {
   state->flagsPF[7] = (StateField){"flags.waterHResp",    FT_INT, &modelFlags.waterHResp,    0};
   state->flagsPF[8] = (StateField){"flags.nitrogenCycle", FT_INT, &modelFlags.nitrogenCycle, 0};
   state->flagsPF[9] = (StateField){"flags.anaerobic",     FT_INT, &modelFlags.anaerobic,     0};
-  state->flagsPF[10] = (StateField){"flags.invalid",      FT_INVALID, NULL, FIELD_INVALID};
+  state->flagsPF[10] = (StateField){"flags.flooding",     FT_INT, &modelFlags.flooding},
```
[the remark this answers is not in the document]
Comment (Alomir) [reply src/sipnet/restart.c]: Good catch

File: src/sipnet/restart.c [hunk 10]
```diff
@@ -160,7 +161,8 @@ void initResetState(RestartState *state, MeanTracker *npp) {
   state->flagsPF[7] = (StateField){"flags.waterHResp",    FT_INT, &modelFlags.waterHResp,    0};
   state->flagsPF[8] = (StateField){"flags.nitrogenCycle", FT_INT, &modelFlags.nitrogenCycle, 0};
   state->flagsPF[9] = (StateField){"flags.anaerobic",     FT_INT, &modelFlags.anaerobic,     0};
-  state->flagsPF[10] = (StateField){"flags.invalid",      FT_INVALID, NULL, FIELD_INVALID};
+  state->flagsPF[10] = (StateField){"flags.flooding",     FT_INT, &modelFlags.flooding,      0};
```
Comment (dlebauer) [inline src/sipnet/restart.c:164]: do we need to bump schema version? should the need to bump schema version after adding a term be caught by a test?

File: docs/CHANGELOG.md [hunk 11]
```diff
@@ -50,6 +48,7 @@ sections to include in release notes:
 - Nitrogen demand, fixation flux, and uptake flux (#265)
 - Methane production option, controlled by the `anaerobic` cli option (#269)
 - Nitrogen limitation support (#298)
+- `flooding` cli option and`waterDrainFrac` parameter added (#307)
```
Comment (dlebauer) [inline docs/CHANGELOG.md]:
```suggestion
- `flooding` cli option and `waterDrainFrac` parameter added (#273, #307)
```

File: docs/parameters.md [hunk 12]
```diff
@@ -210,24 +210,25 @@ pool stoichiometry.
...
+| $r_{s,1}$                       | rSoilConst1         |                                                                                                                              | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1) , where W1 = (water/soilWHC)                                                                                                      |
+| $r_{s,2}$                       | rSoilConst2         |                                                                                                                              | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1) , where W1 = (water/soilWHC)                                                                                                      |
```
Comment (dlebauer) [inline docs/parameters.md]:
```suggestion
| $r_{s,1}$                       | rSoilConst1         |                                                                                                                              | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1) , where W1 = clip(water/soilWHC,0,1)                                                                                                      |
| $r_{s,2}$                       | rSoilConst2         |                                                                                                                              | unitless                                       | soil resistance = e^(rSoilConst1 - rSoilConst2 \* W1) , where W1 = clip(water/soilWHC,0,1)                                                                                                      |
```

File: docs/model-structure.md (hunk 5 above)
Comment (dlebauer) [reply docs/model-structure.md:756]: is 1/dt still needed?

**Review by dlebauer — COMMENTED** [review]
Documentation suggestions added

File: docs/parameters.md [hunk 13]
```diff
@@ -210,24 +210,25 @@ pool stoichiometry.
...
+| $f_{\text{fastflow}}$           | fastFlowFrac        | fraction of water entering soil that goes directly to drainage                                                               | unitless                                       |                                                                                                                                                                                         |
+| $f_{\text{drain}}$              | waterDrainFrac      | fraction of excess soil water that drains per day                                                                            | $\text{day}^{-1}$                              |                                                                                                                                                                                         |
```
Comment (dlebauer) [inline docs/parameters.md:222]:
```suggestion
| $f_{\text{drain}}$              | waterDrainFrac      | fraction of excess soil water that drains per day. Converted internally to a per-timestep drainage flux.                                                                            | $\text{day}^{-1}$                              |                                                                                                                                                                                         |
```

File: docs/model-structure.md [hunk 14]
```diff
@@ -737,18 +737,29 @@ irrigation) is assumed to evaporate the same day and therefore never enters $W_{
...
+If `FLOODING` is enabled, the rate is controlled by the drainage parameter $f_{\text{drain}}$, defined as the fraction
+of excess soil water that can be removed per day.
 
 \begin{equation}
 F^W_{\text{drainage}} = f_\text{drain} \cdot \max(W_{\text{soil}} - W_{\text{WHC}}, 0)
-\label{eq:drainage}
+\label{eq:drainage_flooding}
 \end{equation}
 
+Flooding can be simulated by using a combination of low $f_{\text{drain}}$ and sufficient
```
Comment (dlebauer) [inline docs/model-structure.md:760]:
```suggestion
The parameter $f_{\text{drain}}$ is a per-day rate ($\text{day}^{-1}$) and is internally converted to a per-timestep flux using the model timestep length $\Delta t$.

Flooding can be simulated by using a combination of low $f_{\text{drain}}$ and sufficient 
```

File: src/sipnet/sipnet.c [hunk 15]
```diff
@@ -977,7 +995,12 @@ void calcSoilWaterFluxes(double *fastFlow, double *evaporation,
 
   // drain any water that remains beyond water holding capacity:
   if (waterRemaining > params.soilWHC) {
-    *drainage = (waterRemaining - params.soilWHC) / (climate->length);
+    double excessWater = waterRemaining - params.soilWHC;
+    if (ctx.flooding) {
+      *drainage = excessWater * params.waterDrainFrac;
```
Comment (dlebauer) [inline src/sipnet/sipnet.c:1000]: Should this also be scaled to per-tilmestep?

```suggestion
      *drainage = excessWater * params.waterDrainFrac / climate->length;
```

**Review by dlebauer — CHANGES_REQUESTED** [review]
I made suggestions assuming that the goal is to use a daily time step and convert it internally; trying to keep code and docs consistent with this and with each other, and clear. I also unresolved a few suggestions that I think shouldn't have been resolved, but perhaps were through the many code changes.

**Comment by dlebauer** [comment]
Note: added the alternate flooding scheme ideas to .alternate-model-structure.md #313

File: src/sipnet/restart.c (hunk 10 above)
Comment (Alomir) [reply src/sipnet/restart.c:164]: That's a really good question; TBH, I've been ignoring schema version for now, thinking of this as still "in flux". I think once we release a new SIPNET version, we should probably start bumping the schema version.

File: docs/model-structure.md (hunk 5 above)
Comment (Alomir) [reply docs/model-structure.md:756]: Not with our current definition of f_drain

File: docs/model-structure.md (hunk 14 above)
Comment (Alomir) [reply docs/model-structure.md:760]: All fluxes are per-day - which is converted to a per-time-step _amount_ when the pools are updated. As f_drain is already a per-day rate, there's no need to apply the climate length here.

File: docs/parameters.md (hunk 13 above)
Comment (Alomir) [reply docs/parameters.md:222]: As in other replies, this is not needed.

File: docs/parameters.md (hunk 12 above)
Comment (Alomir) [reply docs/parameters.md]: Good catch!

File: src/sipnet/sipnet.c (hunk 15 above)
Comment (Alomir) [reply src/sipnet/sipnet.c:1000]: Nope, waterDrainFrac is already a per-day rate.

**Review by dlebauer — APPROVED** [review]
